Mathlib/Tactic/StacksAttribute.lean

```diff
@@ -0,0 +1,117 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import Lean.Elab.Command
+
+/-!
+# The `stacks` attribute
+
+This allows tagging of mathlib lemmas with the corresponding
+[Tags](https://stacks.math.columbia.edu/tags) from the Stacks Project.
+-/
+
+open Lean
+
+namespace Mathlib.Stacks
+
+/-- `Tag` is the structure that carries the data to check if a declaration or an
+import are meant to exist somewhere in Mathlib. -/
+structure Tag where
+  /-- The Stacks Project tag. -/
+  tag : String
+  /-- The name of the declaration with the given tag. -/
+  declName : Name
+  deriving BEq, Hashable
+
+/-- Defines the `tagExt` extension for adding a `HashSet` of `Tag`s
+to the environment. -/
+initialize tagExt : SimplePersistentEnvExtension Tag (HashSet Tag) ←
+  registerSimplePersistentEnvExtension {
+    addImportedFn := fun as => as.foldl HashSet.insertMany {}
+    addEntryFn := .insert
+  }
+
+/--
+`addTagEntry tag declName` takes as input the `String` `tag` and the `Name` `declName` of
+a declaration with the `stacks` attribute.
+It extends the `Tag` environment extension with the data `tag, declName`.
+-/
+def addTagEntry {m : Type → Type} [MonadEnv m] (tag : String) (declName : Name) : m Unit :=
+  modifyEnv (tagExt.addEntry · { tag := tag, declName := declName })
+
+/--
+The syntax for a Stacks tag: it is an optional number followed by an optional identifier.
+This allows `044Q3` and `GH3F6` as possibilities.
+-/
+declare_syntax_cat stackTag
+
+@[inherit_doc Parser.Category.stackTag]
+syntax (num)? (ident)? : stackTag
+
+/-- The `stacks` attribute.
+Use it as `@[stacks TAG "Optional comment"]`.
+The `TAG` is mandatory.
+
+See the [Tags page](https://stacks.math.columbia.edu/tags) in the Stacks project for more details.
+-/
+syntax (name := stacks) "stacks " (stackTag)? (ppSpace str)? : attr
+
+initialize stacksTagAttribute : TagAttribute ←
```

Comment:
Do we need this?

Reply:
No, this is a left-over from a previous attempt: I removed it!